enhancements/network/cloud-egress-ip.md

```diff
@@ -0,0 +1,445 @@
+[4]: https://github.com/kubernetes/cloud-provider-gcp
+[5]: https://github.com/kubernetes-sigs/cloud-provider-azure
+
+## Motivation
+
+Customers would like to be able to run workloads on conventional cloud platforms
+and assign them egress IPs, to be used as traffic "identifiers". The reasons for
+this is usually within the scope of firewall-ing traffic within their corporate
+network which is coming from the cloud, based on the source IP. They would like
+the ability to filter pod egress traffic based upon an allow-list they define,
+and then potentially tie all those allowed-to-egress services to an external IP
+on a dedicated node that they have created in the cloud.
+
+### Goals
+
+For now:
+
+- clearing out the questions in [Summary](#summary)
+
+- Designing a solution to the problem of managing additional private IP address
+  assignments to cloud VMs, in a cloud provider agnostic way.
+
+[#Summary]: #Summary
+
+### Non-Goals
+
+For now:
+
+- Assignment of external IP addresses to cloud VMs
+
+- Assignment of additional NICs to cloud VMs
+
+- Assignment of additional private IP address to any NIC besides the one the
+  network plugin manages.
+
+These three "Non-goals" could become "Goals" if we decide to do a generic "Cloud
+API operator" and write an enhancement proposal for that.  
+
+- Providing a cloud provider solution to OpenShift's networking feature: egress
+  router
+
+## Proposal
+
+### Implementation Details/Notes/Constraints
+
+#### Cloud provider data model for additional private IP assignment
+
+The following section describes initial researched information w.r.t. the API
+for private IP address assignment to VMs, on the three major clouds (AWS, GCP
+and Azure). The goal of this document is not to delve into the details of this,
+as the implementation will have to consider all the specifics. Instead it
+presents the general concepts as to give the reader an understanding of what
+will be required by this future component. The API for doing additional IP
+address assignment to VMs differ slightly between the clouds, but remain quite
+similar in the grand scheme of things.  
+
+##### Azure
+
+Azure requirement for assigning additional private IPs to an already existing
+NIC is detailed in [6]. Although the go SDK for communicating with the Azure API
+has not been studied yet, it is expected to resemble the Azure CLI API:
+
+```
+az network nic ip-config create \
+--resource-group $RG_NAME \
+--nic-name $NIC_NAME \
+--private-ip-address 10.0.0.6 \
+--name $name
+```
+
+Where `$RG_NAME` and `$NIC_NAME` needs to be discovered beforehand when
+assigning an additional IP. It is thus expected that this component will be able
+to retrieve such information prior to performing the IP assignment request to
+the Azure API. Note: the Azure doc does not make any particular reference to
+IPv4 / IPv6, so it's not clear whether both are supported. Azure does however
+claim to be a "dual-stack" cloud, so it is assumed that both are supported.
```

Comment:
I wouldn't assume that. The API may make different assumptions about IPv6 use cases

Reply:
OK, removing that part.